### PR TITLE
Add the Swift equivalent of the llvm.assume intrinsic

### PR DESCRIPTION
**Pull request by aschwaighofer**
rdar://44947459

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Review by stephentyrone — APPROVED** [review]
LGTM, thanks!

**Comment by aschwaighofer** [comment]
This might be useful to you. The assume intrinsic has side-effects. if this causes issues we have to do something more involved. e.g lowering unreachable control flow to this intrinsic at the end of the swift pipeline (and hoping that llvm's assume side effects are not the problem).

**Comment by milseman** [comment]
Should `_sanityCheck()` use this?

**Comment by stephentyrone** [comment]
@milseman Have we documented what we believe the semantics of _sanityCheck are (or what we want them to be)?

File: stdlib/public/core/Builtin.swift [hunk 1]
```diff
@@ -290,6 +290,11 @@ public func _onFastPath() {
   Builtin.onFastPath()
 }
 
+@_transparent
+public func _assume(_ condition: Bool) {
```
Comment (lorentey) [inline stdlib/public/core/Builtin.swift]: I think this should be `@usableFromInline @_transparent inline` for now. (I don't recall an immediate use case for it in the overlays, and lacking that there is no reason to make it fully public.)

Comment (lorentey) [reply stdlib/public/core/Builtin.swift]: Regarding the name: shouldn't we call attention to the fact this provides no checks whatsoever that the condition holds? `_assume` on its own seems attractively succinct, but this seems like a dangerous tool with a rather narrow use case. IIUC, this isn't a replacement for `_sanityCheck`, for example.

`_uncheckedAssume(_:)` might be a better name. Arguably we should stick an `unsafe` in there too. 

`_blindlyTrust(that:)`? 

`_weHoldThisTruthToBeSelfEvident(_:)`??

File: test/IRGen/assume.swift [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// CHECK-LABEL: define swiftcc i64 @"$s1A10testAssumeyS2iF"(i64)
+// CHECK:   [[COND:%.*]] = icmp sgt i64 %0, -1
+// CHECK:   tail call void @llvm.assume(i1 [[COND]])
+// CHECK:   [[RES:%.*]] = lshr i64 %0, 6
+// CHECK:   ret i64 [[RES]]
+
+public func testAssume(_ i: Int) -> Int {
+  let cond = i >= 0
+  _assume(cond)
+  return i / 64
```
Comment (lorentey) [inline test/IRGen/assume.swift]: Sweet!

**Review by lorentey — APPROVED** [review]
Nice! This will come handy. I don't like the name of the stdlib symbol though -- it should be a lot scarier. Make me think twice before using it!

File: stdlib/public/core/Builtin.swift (hunk 1 above)
Comment (jrose-apple) [reply stdlib/public/core/Builtin.swift]: It's clearly not self-evident since otherwise the compiler would be able to infer it. ;-)

Comment (stephentyrone) [reply stdlib/public/core/Builtin.swift]: `_assume` has the virtue that it maps directly to the LLVM builtin name. That's worth something. I would be ok with `_uncheckedAssume`.

`_iSolemlySwearThatIAmUpToNoGood`.

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
@swift-ci Please test

**Comment by aschwaighofer** [comment]
My latest version includes ``@usableFromInline`` as requested. I don't think this is necessary, ``@_transparent`` should force inlining.

```swift
// Optimizier hint that the condition is true. The condition is unchecked.
// The builtin acts as an opaque instruction with side-effects.
@usableFromInline @_transparent
func _uncheckedUnsafeAssume(_ condition: Bool) {
   _ = Builtin.assume_Int1(condition._value)
 }

```

**Comment by jrose-apple** [comment]
`_transparent` forces inlining when there's a direct call, but nothing stops someone from taking a reference to `_uncheckedUnsafeAssume`. That's why the `usableFromInline` is necessary.

**Comment by aschwaighofer** [comment]
I see.

File: stdlib/public/core/Builtin.swift [hunk 3]
```diff
@@ -290,6 +290,13 @@ public func _onFastPath() {
   Builtin.onFastPath()
 }
 
+// Optimizier hint that the condition is true. The condition is unchecked.
```
Comment (harlanhaskins) [inline stdlib/public/core/Builtin.swift]: Nit: `Optimizer`

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test and merge

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test and merge

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test mac os

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test mac os

**Comment by aschwaighofer** [comment]
@swift-ci Please smoke test os x

**Pull request merged**